The localization of Chinese has been completed

maikebing opened the pull request:
* Extract all strings and WPF strings that need to be translated And the use of TomEnglert.ResXManager for translation work, of which Chinese has been translated.

maikebing commented:
@codecadwallader   do  you have time  merge it ?

codecadwallader commented:
I wasn't sure if you were ready for review, I will look over it.

codecadwallader commented on CodeMaid/StringResourceKey.cs:
```diff
@@ -0,0 +1,34 @@
+
+namespace SteveCadwallader.CodeMaid
+{
+    internal class StringResourceKey : Properties.Resources
```
I think I know what's going on here, but can you explain this file to me to make sure?

maikebing replied:
Stringresourcekey is the name of the string resource class that is used by default when Tom-englert/resxresourcemanager is automatically generated. It's not perfect, but it's the best localized translation tool I've ever seen. Use of this file is still to cater to the tool to make the translation faster completion.

codecadwallader replied:
Ok, but can you explain what this file is doing to me?

maikebing replied:
Using resources requires referencing namespace SteveCadwallader.CodeMaid.Properties, but using Stringresourcekey does Not require Stevecadwallader.codemaid, because the root namespace is it. It would be awkward to add namespaces to all required code SteveCadwallader.CodeMaid.Properties.

codecadwallader replied:
It looks like this class isn't doing anything at all anymore, so we could delete it and have everything point consistently to Resources.  I know it means adding a using statement to do those C# files, but I do not think that is a bad thing and I think it's simpler than having two seeming versions of the resources.

codecadwallader commented on CodeMaid/UI/Dialogs/CleanupProgress/CleanupProgressViewModel.cs:
```diff
@@ -176,7 +176,7 @@ private void backgroundWorker_ProgressChanged(object sender, ProgressChangedEven
 
             CountProgress = currentCount;
-            CurrentFileName = currentItem.Name;
+            CurrentFileName = string.Format(StringResourceKey.Cleaning0, currentItem.Name);
```
This isn't the current file name anymore, but an output string.  I think this was done to avoid using the StringFormat in XAML.  If possible it would be preferable to keep the StringFormat in XAML, since this class is a ViewModel.

maikebing replied:
I tried the XAML stringformat, but it doesn't seem to work.

codecadwallader replied:
Ok, another approach would be to split the label into two components then.  One for the "Cleaning" which can be translated directly and one for the CurrentFileName.  That would allow this ViewModel to be reverted to not having View code in it.

codecadwallader commented on CodeMaid/UI/Dialogs/Options/Cleaning/CleaningFileTypesDataTemplate.xaml:
```diff
@@ -1,6 +1,7 @@
 <ResourceDictionary xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
         xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
         xmlns:local="clr-namespace:SteveCadwallader.CodeMaid.UI.Dialogs.Options.Cleaning"
+        xmlns:properties="clr-namespace:SteveCadwallader.CodeMaid.Properties"
```
In other files the convention is `xmlns:p` so it would be good to stick to that here too.

maikebing replied:
'properties' are the names that Tom-englert/resxresourcemanager use by default, so they are still for faster translation.

codecadwallader replied:
I understand a lot of the code was auto-generated by a tool, but we've been really diligent about trying to have as clean of a code base as possible.  To accept the pull request the auto-generated code will have to be updated to match conventions of the existing code base.

maikebing replied:
has been modified ！
91af8fdb442cd0bae3bd6ce5ed563fd5d394c227

codecadwallader commented on CodeMaid/UI/Dialogs/Options/OptionsViewModel.cs:
```diff
@@ -217,10 +217,10 @@ private void OnExportCommandExecuted(object parameter)
                 FileName = "CodeMaid",
                 DefaultExt = ".config",
-                Filter = "Config files (*.config)|*.config|All Files (*.*)|*.*"
+                Filter = StringResourceKey.OptionsViewModel_OnImportCommandExecuted_ConfigFilesConfigConfigAllFiles
```
This auto-generated name is very awkward.. perhaps ConfigFileFilter would be clearer.

maikebing replied:
Yes, it's very long, I submit a refactoring code?

codecadwallader replied:
That would be great, thanks.

maikebing replied:
e279ed8

maikebing commented on CodeMaid/CodeMaid.csproj:
```diff
@@ -791,13 +797,22 @@
       <MergeWithCTO>true</MergeWithCTO>
       <ManifestResourceName>VSPackage</ManifestResourceName>
     </EmbeddedResource>
+    <EmbeddedResource Include="source.extension.zh-Hans.resx">
```
This is my mistake, not noticing this problem.

maikebing replied:
Visual Studio compiles automatic local language using different manifest files

maikebing commented on CodeMaid/CodeMaid.vsct:
```diff
@@ -288,13 +288,15 @@
           If you do not want an image next to your command, remove the Icon node or set it to <Icon guid="guidOfficeIcon" id="msotcidNoIcon" /> -->
       <Button guid="GuidCodeMaidCommandAbout" id="CmdIDCodeMaidAbout" priority="0x0100" type="Button">
         <Icon guid="GuidImageInfo" id="IconInfo" />
+        <CommandFlag>TextChanges</CommandFlag>
```
Except for the dynamic text, I didn't find a better way to do it.There seems to be an article that says VSCT can have multiple languages. But I don't know how to do it so far.

codecadwallader replied:
I think looking into this further will be necessary.. in the past the performance hit was debilitating as our extension as well as other ones would cause each other to infinitely regenerate the text.

maikebing replied:
I tried the method described in MSDN, but I didn't succeed and I didn't know what the problem was. 
url: https://msdn.microsoft.com/en-us/library/ee943168.aspx

maikebing replied:
3642dec9c02dbf92e4748993daa093df86fa7d0b

maikebing commented on CodeMaid/Integration/Commands/CleanupAllCodeCommand.cs:
```diff
@@ -48,12 +48,12 @@ protected override void OnExecute()
             {
-                MessageBox.Show(@"Cleanup cannot run while debugging.",
-                                @"CodeMaid: Cleanup All Code",
+                MessageBox.Show(StringResourceKey.CleanupAllCodeCommand_OnExecute_CleanupCannotRunWhileDebugging,
```
I used tom-englert/ResXResourceManager to achieve string extraction, and I used the variable name that it automatically generated.In order to complete the extraction of string statements as soon as possible, I have no refactoring.

codecadwallader replied:
I would like to have these simplified before accepting the pull request.. otherwise the code will be pretty difficult to read.

maikebing replied:
Code has been refactored e279ed86150b93d1fb93cc5d6af641458c18c6ee

maikebing commented on CodeMaid/Integration/Commands/CommentFormatCommand.cs:
```diff
@@ -27,7 +27,7 @@ internal CommentFormatCommand(CodeMaidPackage package)
                    new CommandID(PackageGuids.GuidCodeMaidCommandCommentFormat, PackageIds.CmdIDCodeMaidCommentFormat))
         {
-            _undoTransactionHelper = new UndoTransactionHelper(package, "CodeMaid Format Comment");
+            _undoTransactionHelper = new UndoTransactionHelper(package, StringResourceKey.CommentFormatCommand_CommentFormatCommand_CodeMaidFormatComment);
```
StringResourceKey is the name of the string resource class that is used by default when tom-englert/ResXResourceManager is automatically generated. It's not perfect, but it's the best localization translation tool I've ever met.

codecadwallader replied:
Sometimes StringResourceKey is used.. sometimes Resources is used.  I think to avoid confusion Resources should always be used where possible.

maikebing replied:
Using resources requires referencing namespace SteveCadwallader.CodeMaid.Properties, but using Stringresourcekey does not require Stevecadwallader.codemaid, Because the root namespace is it. It would be awkward to add namespaces to all the required code SteveCadwallader.CodeMaid.Properties. However, it is easier to refer to SteveCadwallader.CodeMaid.Properties in WPF.

maikebing commented on CodeMaid/Properties/Resources.zh-Hans.resx:
```diff
@@ -0,0 +1,762 @@
+    <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
+  </resheader>
+  <data name="BuildProgressToolWindow_BuildProgress" xml:space="preserve">
+    <value>生成进度</value>
```
The first is through the use of Tom-englert/resxresourcemanager Microsoft translation, and then manual proofreading.，  It conforms to the Chinese language grammar.

codecadwallader replied:
Ok, thanks.

maikebing commented:
Is there anything else you need to change?  @codecadwallader

codecadwallader commented on CodeMaid/CodeMaid.csproj:
```diff
@@ -172,8 +172,8 @@
     </None>
+    <Compile Include="StringResourceKey.cs" />
     <Compile Include="CodeMaid.cs">
-      <AutoGen>True</AutoGen>
```
Since this is an auto-generated file, I think we need this flag?

codecadwallader commented on CodeMaid/Integration/Commands/BaseCommand.cs:
```diff
@@ -46,6 +46,10 @@ protected BaseCommand(CodeMaidPackage package, CommandID id)
         private static void BaseCommand_BeforeQueryStatus(object sender, EventArgs e)
         {
             BaseCommand command = sender as BaseCommand;
+            //if (string.IsNullOrEmpty(command.Text))
```
Please remove commented out code.

maikebing commented:
I have finished. Take the time to review the code!

codecadwallader commented:
Thank you for the code updates.  The code is looking good and I have no further comments. :thumbsup:

I pulled down the code and tried to run it locally, but when I startup an experimental instance of Visual Studio the CodeMaid menu is not present at all.  I suspect the changes to the .vsct or source.extension files have caused an issue.  Can you please look into it?

maikebing commented:
[ResetEnvironment.zip](https://github.com/codecadwallader/codemaid/files/1821518/ResetEnvironment.zip)

codecadwallader commented:
Yeah I tried the usual "Reset the Visual Studio 2017 Experimental Instance" batch file that comes with the SDK but it did not fix the issue.

maikebing commented:
My VS2017 version is 15.6.2. The code I tested under the master branch doesn't seem to be working properly, and I'll find out why.

maikebing commented:
I tested it and looked normal together.
I uninstalled the codemaid and removed all appdatalocalmicrosoftvisualstudio15.0_f203e778exp. The associated registry key is also deleted.
When I am debugging, I can enter the debugging, also can display the Codemaid menu！

here are also files in the directory 
 C:\Users\MaiKeBing\AppData\Local\Microsoft\VisualStudio\15.0_f203e778Exp\Extensions\Steve Cadwallader\CodeMaid\10.4

maikebing commented:
But if I use the parameter/rootsuffix EXP1 then the Codemaid menu does not show

maikebing commented:
Change the parameters anyway, VS2017 will put codemaid here.
C:\Users\MaiKeBing\AppData\Local\Microsoft\VisualStudio\15.0_f203e778Exp

maikebing commented:
If you use other parameters, please try    “/rootsuffix Exp”

codecadwallader commented:
Yep /rootsuffix Exp is baked into the repository: https://github.com/codecadwallader/codemaid/blob/master/CodeMaid/CodeMaid.csproj.user#L6

I'm not sure I follow your comments about what you tried and what happened?

codecadwallader commented:
As a heads up - I just merged another pull request that allows users to turn on/off different features.  It looks like your PR will still be able to merge in cleanly, but there are some new user options that will also need setup for translation.  A good question is how will we perform translations into Chinese going forwards every time the UI changes?  I can use Google Translate, but I know that's not a great source...

maikebing commented:
It is recommended that you use Microsoft's Bing translation tool

maikebing commented:
I like codemaid very much, if need, I can translate at any time.

heku commented:
Hello @maikebing, glade to know you, I'm a Chinese too. Becuase @codecadwallader probably don't know any Chinese, he might reviewed the transalation via some transaltion tools, but as we know, the transalatio tools sometimes don't give us a good(or correct) answers, do you mind I review your Chinese transalation on behalf of codecadwallader? I'd like to work with you to see if we could make it better. Thank you.

maikebing commented:
@heku  Yes, you can.

heku commented:
From my overall review, I have some questions, for better communication, I speak Chinese here.
你好maikebing, 我拉了你最新的localizaiton分支看了下，感觉有以下几个小问题需要跟你确认一下：
1. 我调试的时候看不到任何菜单（VS2017最新版），你那边可以？
2. source.extension.vsixmanifes这个文件没有本地化，你可以参考我的两个项目中的做法。
3. 我感觉英文版本的resx文件可能不用两个（一个xx.resx；一个xx.en-US.resx），应该只要一个默认的就行了。
4. vsct英文版本只保留CodeMaid.en-US.vsct的就行了（CodeMaid.vsct不需要了）。
5. 部分翻译我觉得我们最好不直接字面翻译，像“spade”翻译成“码锹”，"digging"翻译成"发掘"...用这些词感觉很奇怪，难理解是什么意思。我们要重新想想，从功能的角度上来想个词来描述这些功能，让用户一看就知道是什么。
6. 还有一些关键字类型的英文单词，我建议还是直接保留关键字，比如region，attribute，因为用户都是开发人员，这样肯定没有歧义。当前发现region翻译成“大纲”，不太准确（VS把outline视为大纲），attribute翻译成”属性“感觉与property的歧义，或者用”特性“或者直接用attribute。

以上是一些个人意见，今天只是大体看了下，后面有时间我会继续看细一点。我问题有点多，请勿介意，我只是想和你一起努力让CodeMaid更好一点。

maikebing commented:
> 我调试的时候看不到任何菜单（VS2017最新版），你那边可以？
我这里没有问题， 最新版本。 

> source.extension.vsixmanifes这个文件没有本地化，你可以参考我的两个项目中的做法。
这个本地化只是资源文件
source.extension.en-US.resx	
source.extension.resx	
source.extension.zh-Hans.resx



> 我感觉英文版本的resx文件可能不用两个（一个xx.resx；一个xx.en-US.resx），应该只要一个默认的就行了。
一个是中间语言版本，保持VS自动生成的就好。 


> vsct英文版本只保留CodeMaid.en-US.vsct的就行了（CodeMaid.vsct不需要了）。
CodeMaid.vsct
本地化后 它在这里只是一个文件，不具备任何编译意义 ， 没有合并之前不想删除它，目的是便于同步翻译。 其他语言以它为蓝本，包括英文。 

> 部分翻译我觉得我们最好不直接字面翻译，像“spade”翻译成“码锹”，"digging"翻译成"发掘"...用这些词感觉很奇怪，难理解是什么意思。我们要重新想想，从功能的角度上来想个词来描述这些功能，让用户一看就知道是什么。
spade 在英文中可以翻译为铁铲 、铁锹的意思， 我觉得应该把它作为一个工具名称翻译， 而不是功能名称翻译。 所以， 我觉得既然是 代码方面的铁锹， 那应该翻译为  码锹 ， 处理代码的铁锹。 
digging 在英文中有挖掘的意思， 直接翻译为挖掘肯定不合适， 我觉得带有发现 、寻找、深入寻找的意思， 并且鉴于 spade   一次， 我想 “发掘” 更合适一点点。 


> 还有一些关键字类型的英文单词，我建议还是直接保留关键字，比如region，attribute，因为用户都是开发人员，这样肯定没有歧义。当前发现region翻译成“大纲”，不太准确（VS把outline视为大纲），attribute翻译成”属性“感觉与property的歧义，或者用”特性“或者直接用attribute。

关于 region 
msdn中
   
  #region请参见  
  C#   预处理器指令  
  #region   使您得以在使用   Visual   Studio   代码编辑器的大纲显示功能时指定可展开或折叠的代码块。

maikebing commented:
@codecadwallader   French, German and other languages may require you to find someone to translate, I hope localization more than Chinese.

heku commented:
@maikebing Sorry for the late response, I'm a little busy these days. I'm thinking to find a better way to help us organize the localization code more cleaner without duplicated files (such as `CodeMaid.vsct`/`CodeMaid.en-US.vsct` and other `*.resx` files). I know the `CodeMaid.vsct` file was there for 'Extensibility Tools' to export IDs and the `Resources.resx` file was there for generating a helper class to get the resource strings. I'm glad that I find out some tips might help us.
* For the `CodeMaid.vsct`/`CodeMaid.en-US.vsct` files, we can let the `CodeMaid.vsct` file contains only non-menu elements (IDs, groups, keybindings etc), all UI related elements are put into `CodeMaid.Culture.vsct` files, in each culture version file, add `<Include href="CodeMaid.vsct" />`. In this way, we can both keep the `CodeMaid.vsct` file (for generating help class) and remove most duplicated content. 
* For the `*.resx` files, I think it's better to re-use the `source.extension.xx.resx` rather than adding several new resource files, I believe you added new resource files is because you want a helper class to access the resource strings, as VS cannot generate that helper class for a resource file with culture name. I think we can add a T4 file to do that by reading the `source.extension.en-US.resx` to generate a helper class. Then we can merge the `Properties/Resources.xxx.resx` into `source.extension.xxx.resx`.

I have done these in my repo [Kool.VsDiff](https://github.com/heku/Kool.VsDiff), you take a review for reference.  

@codecadwallader How do you think?

codecadwallader commented:
Thank you for stepping in and helping with this pull request.

- I think it would be very helpful to remove duplicated code in the CodeMaid.vsct files so yes if we can just include the the core reference that would be great.

- There were already two separate resource files before these changes.  The source.extension.resx just holds 3 values and was items needed to build the manifest only.  The Resources.resx is the primary one that holds everything else.  I think this is OK as-is and moving all of the values and adding a T4 template isn't necessary.

Thank you again!

maikebing commented:
@codecadwallader @heku

heku commented:
Hi @codecadwallader , these days @maikebing and I work together to enhance the Chinese localization, we have a question need your help that we don't have any good idea on how to translate the 'Spade', it's literal meaning is
![th](https://user-images.githubusercontent.com/3367199/38463277-8dec74e0-3b29-11e8-85c4-91b8f01ceb60.jpg)

If we translate it to the name of the tool in Chinese, the word might look quite confused for users. I'm wondering if the 'Spade' you chose has any other meaning in your side/culture. If yes, what's that? If no, I prefer translate it from it's functionality as 'Code Outline' in Chinese, how do you think?

codecadwallader commented:
@heku thank you for working on it together, I appreciate it. :)

You have the correct meaning of Spade, it is a digging tool like a shovel.  We call that visualization feature "digging" because it lets you "dig" through your source code quickly without having to scroll through every line of code.  The name "Spade" was added to it as another word for digging because it rhymes with CodeMaid in English.  It makes sense that rhyme will not carry over when translated.  How about translating "CodeMaid Digging" instead?

heku commented:
@codecadwallader Thank you for confirmation. Honestly, I think it's still not easy to give it an easily understood name in Chinese. In my above comment I suggested naming it as 'Code Outline', because the VS has a tool window called 'Document Outline', which is similar with spade tool window.

maikebing commented:
If CodeMaid is a maid of code, "Spade" has been used in English,I think the direct translation as "A spade for digging code", So it's called 码锹（ Code Spade）， I think it will be a good name.!

heku commented:
OK, let us use this translation at this stage. I suggest for every localization, open an issue for any feedback and discussion with a title in two languages. For example, for current Chinese localization, it can be “Chinese localization feedback/discussion/suggestion 中文本地化 反馈/讨论/建议”. how do you guys think?

codecadwallader commented:
Ok, to make sure I understand correctly you're saying that each time there is a new feature with new translations needed, new issues should be created for every localization where translation is needed?  That seems fine to me.

heku commented:
Basically Yes, my thought was openning one issue with a title in two language (en+local lang) for every language/localization for any feedback/suggestion. But now I think this might not be a good idea as people always like to create a new issue rather than comment on an existing the issue.

maikebing commented:
@heku  @codecadwallader    Is there any question about this?

maikebing commented:
@heku  @codecadwallader    Is there any question about this?

heku commented:
No, I'm fine.

maikebing commented:
source.extension.resx     is automatically generated each time !

codecadwallader commented:
Ok great!  Thanks both (especially @maikebing ) for your hard work on this.  I will merge it in now and we can review/test from the mainline. :)

codecadwallader commented:
@heku it looks like some of your .csproj cleanups may have gotten merged out when this got merged into master.  It looks like @maikebing did do a merge down from master into this PR branch so I'm not exactly sure why that happened, but perhaps it was too many changes in the same place for Git to resolve.

heku commented:
@codecadwallader I'll have it a check on your latest code, sorry for the late response as I'm quite busy these days.
